src/metaprob/autodiff.cljc

```diff
@@ -0,0 +1,247 @@
+(ns metaprob.autodiff
+  (:refer-clojure :exclude [+ * - / == < > <= >= zero? number? pos? neg?]))
+
+;; Dual numbers represent computations dependent on a value with respect to which we are differentiating.
+;; derivative contains the current derivative; value contains the current value; and tag is a special tag
+;; that helps distinguish between multiple "levels" of nested differentiations.
+(defn make-dual-number [tag value derivative]
+  {:tag tag :value value :derivative derivative})
+
+(declare * + - /)
+
+(def non-dual? clojure.core/number?)
+
+;; Take a unary operation f on numbers
+;; and lift it to work with dual numbers.
+;; Note that df-dx should itself be a lifted
+;; function for computing the derivative of f
+;; at a point.
+(defn lift-real->real
+  [f df-dx]
+  (fn new-f [x]
+    (if (non-dual? x)
+      (f x)
+      (make-dual-number (:tag x)
+                        (new-f (:value x))
+                        (* (df-dx (:value x)) (:derivative x))))))
+
+;; Lift a binary operation on numbers to work
+;; with dual numbers. We have to consider several
```

Comment:
Could you please describe these use cases and what role the comparison of keys plays in ensuring correctness of the autodiff, if any?

Reply:
The goal of `new-f` is to return a new value such that

- if `x1` and `x2` were both regular numbers, so is the new value `(f x1 x2)`
- if either one was dual (i.e., was a function of some variable with respect to which this computation is being differentiated), return a new dual number with an updated value and derivative

So, at the most basic level, the cases are checking which of `x1` and `x2` are numbers vs. dual numbers. If `x1` is a dual number and `x2` is a number, we need to use `df/dx1` to update the derivative. Similarly, if `x2` is a dual number but `x1` is a regular number, we need to use `df/dx2`. If they are both dual numbers, they could be from "the same differentiation" (the same tag), or different. If they are different, we handle the highest tag ("innermost" differentiation) first, and let the recursive call handle the rest. If they are the same (the last case), we need to sum the effects of both derivatives to correctly apply the chain rule.

Reply:
I see, thanks.  So it sounds like:

1. The "dual number" data structure here is a tree representing a partially completed application of autodiff.
2. There is a notion of tree weight for such a data structure: roughly, it's the number of deferred differentiation operations.  A tree weight weight 0 is a `(value, derivative)` pair, where both `value` and `derivative` are bare floats.
3. Each of the "lifted" operators operates on such a tree and, if the tree weight is positive, decreases the tree weight by at least 1.

Is that right?

Reply:
I don't think that's quite right.
1. In what sense are you thinking that it is "partially completed"?  A dual number always represents some concrete value `v(a)` dependent on a variable `x` at `x=a`, and its derivative `(dv/dx)(a)`. When `v(a)` is itself a dual number, the bare float we take it to represent can be computed by iterating the `(:value ...)` function until you get a bare float. Similarly, when `(dv/dx)(a)` is a dual number, you can get the value represented by that dual number by iterating `(:value ...)` until you get a bare float.

2. Sure, you could define such a notion, though I'm not sure how exactly I'd interpret it (there's probably a good way!). Note that the `:tag` is not the tree weight in any sense.

3. Lifted operators leave the tree structure completely unchanged. They ensure that every part of the tree is updated to reflect the fact that the function `v` has changed (it has been composed with whatever operator is being lifted).

Reply:
Ah ok, sounds like I didn't get it right.  Based on this line:

>  If they are different, we handle the highest tag ("innermost" differentiation) first, and let the recursive call handle the rest.

it sounds like the data structure is coupled to a specific way of computing derivatives, and because we rely on recursive calls to "handle the rest", it sounds like that algorithm involves a loop.  And in order to reason about the loop (e.g. guaranteeing it terminates), I was trying to come up with an invariant that holds for each iteration of the loop.  It seems like there should be one, though perhaps we rely on the Clojure interpreter's call stack rather than reifying it explicitly in a tree data structure.

How much of the above made sense?